File: api/activation.js
/**
 * Email activation after password registration.
 */
const crypto = require("crypto");
const { emailConfigured, publicAppUrl, sendMail } = require("./mail");

const ACTIVATION_DETAIL =
  "Account is not activated. Check your email and click the activation link.";
const REGISTERED_DETAIL = "Check your email to activate your account before logging in.";
const RESEND_DETAIL = "If that email needs activation, we sent a new link.";
const INVALID_LINK = "This activation link is invalid or has expired.";
const ACTIVATED_DETAIL = "Account activated. You can log in now.";

function activationHours() {
  const n = Number(process.env.ACTIVATION_TOKEN_HOURS || 48);
  return Number.isFinite(n) && n >= 1 ? n : 48;
}

function isActive(user) {
  if (!user) return false;
  if (user.is_active === undefined || user.is_active === null) return true;
  return Boolean(user.is_active);
}

function issueActivationToken(db, userId) {
  db.prepare("DELETE FROM email_activation_tokens WHERE user_id = ?").run(userId);
  const key = crypto.randomBytes(32).toString("base64url");
  const expires = new Date(Date.now() + activationHours() * 3600 * 1000).toISOString();
  db.prepare(
    "INSERT INTO email_activation_tokens (key, user_id, expires_at) VALUES (?, ?, ?)",
  ).run(key, userId, expires);
  return { key, user_id: userId, expires_at: expires };
}

function activationLink(key) {
  return `${publicAppUrl()}/activate?token=${key}`;
}

async function sendActivationEmail(user, token) {
  const link = activationLink(token.key);
  const hours = activationHours();
  const subject = "Activate your Too Many Games account";
  const text =
    `Hi ${user.username},\n\n` +
    `Thanks for signing up. Activate your account by opening this link:\n${link}\n\n` +
    `The link expires in ${hours} hours.\n` +
    `If you did not create this account, you can ignore this email.\n`;
  const html =
    `<p>Hi ${user.username},</p>` +
    `<p>Thanks for signing up. Activate your account by clicking the button below.</p>` +
    `<p><a href="${link}" style="display:inline-block;background:#7c3aed;color:#fff;` +
    `padding:12px 20px;border-radius:999px;text-decoration:none;font-weight:700">` +
    `Activate account</a></p>` +
    `<p>Or paste this link into your browser:<br>${link}</p>` +
    `<p>The link expires in ${hours} hours. If you did not create this account, ignore this email.</p>`;
  return sendMail({ to: user.email, subject, text, html });
}

async function issueAndSendActivation(db, user) {
  if (!emailConfigured()) {
    return { ok: false, error: "Email delivery is not configured.", status: 503 };
  }
  const token = issueActivationToken(db, user.id);
  const sent = await sendActivationEmail(user, token);
  if (!sent.ok) return { ok: false, error: sent.error, status: 500, token };
  return { ok: true, token };
}

function activateWithKey(db, key) {
  const tokenKey = typeof key === "string" ? key.trim() : "";
  if (!tokenKey) {
    return { ok: false, error: INVALID_LINK, status: 400 };
  }
  const row = db
    .prepare("SELECT * FROM email_activation_tokens WHERE key = ?")
    .get(tokenKey);
  if (!row) {
    return { ok: false, error: INVALID_LINK, status: 400 };
  }
  if (new Date(row.expires_at).getTime() < Date.now()) {
    db.prepare("DELETE FROM email_activation_tokens WHERE key = ?").run(tokenKey);
    return { ok: false, error: INVALID_LINK, status: 400 };
  }
  db.prepare("UPDATE users SET is_active = 1 WHERE id = ?").run(row.user_id);
  db.prepare("DELETE FROM email_activation_tokens WHERE user_id = ?").run(row.user_id);
  const user = db.prepare("SELECT * FROM users WHERE id = ?").get(row.user_id);
  return { ok: true, user };
}

function markEmailVerified(db, user) {
  if (!user) return user;
  if (!isActive(user)) {
    db.prepare("UPDATE users SET is_active = 1 WHERE id = ?").run(user.id);
    user.is_active = 1;
  }
  db.prepare("DELETE FROM email_activation_tokens WHERE user_id = ?").run(user.id);
  return user;
}

module.exports = {
  ACTIVATED_DETAIL,
  ACTIVATION_DETAIL,
  INVALID_LINK,
  REGISTERED_DETAIL,
  RESEND_DETAIL,
  activateWithKey,
  activationHours,
  activationLink,
  emailConfigured,
  isActive,
  issueAndSendActivation,
  issueActivationToken,
  markEmailVerified,
  sendActivationEmail,
};

File: backend/apps/accounts/activation.py
"""Email activation after password registration. Social login is pre-verified."""

from __future__ import annotations

import secrets
from datetime import timedelta

from django.conf import settings
from django.core.mail import send_mail
from django.utils import timezone
from rest_framework.exceptions import APIException, ValidationError

from .models import EmailActivationToken, User

ACTIVATION_DETAIL = "Account is not activated. Check your email and click the activation link."
REGISTERED_DETAIL = "Check your email to activate your account before logging in."
RESEND_DETAIL = "If that email needs activation, we sent a new link."
INVALID_LINK = "This activation link is invalid or has expired."
ACTIVATED_DETAIL = "Account activated. You can log in now."


class EmailDeliveryUnavailable(APIException):
    status_code = 503
    default_detail = "Email delivery is not configured."
    default_code = "email_unavailable"


def public_app_url() -> str:
    return (getattr(settings, "PUBLIC_APP_URL", "") or "http://127.0.0.1:3000").rstrip("/")


def activation_hours() -> int:
    try:
        return max(1, int(getattr(settings, "ACTIVATION_TOKEN_HOURS", 48)))
    except (TypeError, ValueError):
        return 48


def email_delivery_configured() -> bool:
    if (getattr(settings, "EMAIL_HOST", "") or "").strip():
        return True
    return bool(getattr(settings, "DEBUG", False))


def require_email_delivery():
    if not email_delivery_configured():
        raise EmailDeliveryUnavailable()


def issue_activation_token(user: User) -> EmailActivationToken:
    EmailActivationToken.objects.filter(user=user).delete()
    return EmailActivationToken.objects.create(
        user=user,
        key=secrets.token_urlsafe(32),
        expires_at=timezone.now() + timedelta(hours=activation_hours()),
    )


def activation_link(token: EmailActivationToken) -> str:
    return f"{public_app_url()}/activate?token={token.key}"


def send_activation_email(user: User, token: EmailActivationToken) -> None:
    link = activation_link(token)
    hours = activation_hours()
    subject = "Activate your Too Many Games account"
    text = (
        f"Hi {user.username},\n\n"
        f"Thanks for signing up. Activate your account by opening this link:\n{link}\n\n"
        f"The link expires in {hours} hours.\n"
        f"If you did not create this account, you can ignore this email.\n"
    )
    html = (
        f"<p>Hi {user.username},</p>"
        f"<p>Thanks for signing up. Activate your account by clicking the button below.</p>"
        f'<p><a href="{link}" style="display:inline-block;background:#7c3aed;color:#fff;'
        f'padding:12px 20px;border-radius:999px;text-decoration:none;font-weight:700">'
        f"Activate account</a></p>"
        f"<p>Or paste this link into your browser:<br>{link}</p>"
        f"<p>The link expires in {hours} hours. If you did not create this account, ignore this email.</p>"
    )
    send_mail(
        subject,
        text,
        getattr(settings, "DEFAULT_FROM_EMAIL", "Too Many Games <noreply@localhost>"),
        [user.email],
        html_message=html,
        fail_silently=False,
    )


def issue_and_send_activation(user: User) -> EmailActivationToken:
    require_email_delivery()
    token = issue_activation_token(user)
    send_activation_email(user, token)
    return token


def activate_with_key(key: str) -> User:
    token_key = (key or "").strip()
    if not token_key:
        raise ValidationError({"detail": INVALID_LINK})
    row = EmailActivationToken.objects.filter(key=token_key).select_related("user").first()
    if not row or row.expires_at < timezone.now():
        if row:
            row.delete()
        raise ValidationError({"detail": INVALID_LINK})
    user = row.user
    user.is_active = True
    user.save(update_fields=["is_active"])
    EmailActivationToken.objects.filter(user=user).delete()
    return user


def mark_email_verified(user: User) -> None:
    """Social providers already verified the address — allow login immediately."""
    changed = []
    if not user.is_active:
        user.is_active = True
        changed.append("is_active")
    if changed:
        user.save(update_fields=changed)
    EmailActivationToken.objects.filter(user=user).delete()
